cloud/blockstore/libs/nbd/netlink_device.cpp

```diff
@@ -0,0 +1,454 @@
+    }
+}
+
+// queries device status and registers callback that will connect
+// or reconfigure (if Reconfigure == true) specified device
+void TNetlinkDevice::ConnectDeviceAsync()
+{
+    TNetlinkSocket nlsocket;
+
+    nl_socket_modify_cb(
+        nlsocket,
+        NL_CB_VALID,
+        NL_CB_CUSTOM,
+        TNetlinkDevice::StatusHandler,
+        this);
+
+    auto* msg = nlmsg_alloc();
+    if (msg == nullptr) {
+        throw TServiceError(E_FAIL) << "unable to allocate netlink message";
+    }
+    genlmsg_put(
+        msg,
+        NL_AUTO_PORT,
+        NL_AUTO_SEQ,
+        nlsocket.Family(),
+        0,
+        0,
+        NBD_CMD_STATUS,
+        0);
+
+    if (DeviceIndex >= 0) {
+        NLA_PUT_U32(msg, NBD_ATTR_INDEX, DeviceIndex);
+    }
+
+    if (nl_send_sync(nlsocket, msg) < 0) {
+        throw TServiceError(E_FAIL)
+            << "failed to configure device, check dmesg";
+    }
+}
+
+int TNetlinkDevice::StatusHandler(nl_msg* msg, void* arg)
+{
+    auto* header = static_cast<genlmsghdr*>(nlmsg_data(nlmsg_hdr(msg)));
+    auto* conn = static_cast<TNetlinkDevice*>(arg);
+    auto Log = conn->Log;
+    nlattr* attr[NBD_ATTR_MAX + 1];
+    nlattr* deviceItem[NBD_DEVICE_ITEM_MAX + 1];
+    nlattr* device[NBD_DEVICE_ATTR_MAX + 1];
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wc99-designator"
+    nla_policy deviceItemPolicy[NBD_DEVICE_ITEM_MAX + 1] = {
+        [NBD_DEVICE_ITEM] = {.type = NLA_NESTED},
+    };
+    nla_policy devicePolicy[NBD_DEVICE_ATTR_MAX + 1] = {
+        [NBD_DEVICE_INDEX] = {.type = NLA_U32},
+        [NBD_DEVICE_CONNECTED] = {.type = NLA_U8},
+    };
+#pragma clang diagnostic pop
+
+    if (nla_parse(
+            attr,
+            NBD_ATTR_MAX,
+            genlmsg_attrdata(header, 0),
+            genlmsg_attrlen(header, 0),
+            NULL))
+    {
+        STORAGE_ERROR("unable to parse NBD_CMD_STATUS response");
+        return NL_OK;
+    }
+
+    if (!attr[NBD_ATTR_DEVICE_LIST]) {
+        STORAGE_ERROR("did not receive NBD_ATTR_DEVICE_LIST");
+        return NL_OK;
+    }
+
+    if (nla_parse_nested(
+            deviceItem,
+            NBD_DEVICE_ITEM_MAX,
+            attr[NBD_ATTR_DEVICE_LIST],
+            deviceItemPolicy))
+    {
+        STORAGE_ERROR("unable to parse NBD_ATTR_DEVICE_LIST");
+        return NL_OK;
+    }
+
+    if (!deviceItem[NBD_DEVICE_ITEM]) {
+        STORAGE_ERROR("did not receive NBD_DEVICE_ITEM");
+        return NL_OK;
+    }
+
+    if (nla_parse_nested(
+            device,
+            NBD_DEVICE_ATTR_MAX,
+            deviceItem[NBD_DEVICE_ITEM],
+            devicePolicy))
+    {
+        STORAGE_ERROR("unable to parse NBD_DEVICE_ITEM");
+        return NL_OK;
+    }
+
+    if (!device[NBD_DEVICE_CONNECTED]) {
+        STORAGE_ERROR("did not receive NBD_DEVICE_CONNECTED");
+        return NL_OK;
+    }
+
+    conn->ConnectDevice(nla_get_u32(device[NBD_DEVICE_CONNECTED]));
```

Comment:
it should be nla_get_u8 according to this https://elixir.bootlin.com/linux/v5.13.16/source/drivers/block/nbd.c#L2221

Reply:
Yeah, you are right. It's also specified in the parse policy